seguro/common/store.py

```diff
@@ -54,89 +83,88 @@ def __init__(
         """
-        if not self.client.bucket_exists(bucket):
-            print(f"Error: Bucket {bucket} does not exist...")
-            return
-
-        self.client.fget_object(bucket, file, filename)
+        return self.client.fget_object(self.bucket, file, filename)
 
-    def put_file(self, filename, file, bucket=S3_BUCKET):
+    def put_file(self, filename, file):
         """Upload local file and store it in the S3Storage.
 
         Arguments:
             filename -- Name of uploaded file in storage
             file     -- Local file that is used
-
-        Keyword arguments:
-            bucket -- Storage bucket that is considered (default "seguro")
         """
-        if not self.client.bucket_exists(bucket):
-            print(f"Error: Bucket {bucket} does not exist...")
-            return
+        return self.client.fput_object(self.bucket, filename, file)
 
-        self.client.fput_object(bucket, filename, file)
+    def remove_file(self, filename):
+        return self.client.remove_object(self.bucket, filename)
 
-    def write_to_file(self, filename, content, bucket=S3_BUCKET):
+    def put_file_contents(self, filename, content):
         """Write to file stored it in the S3Storage.
 
         Arguments:
             filename -- Name of file in storage
             content  -- Content that is written to file
-
-        Keyword arguments:
-            bucket -- Storage bucket that is considered (default "seguro")
         """
-        if not self.client.bucket_exists(bucket):
-            print(f"Error: Bucket {bucket} does not exist...")
-            return
 
-        self.client.put_object(
-            bucket,
+        return self.client.put_object(
+            self.bucket,
             filename,
             io.BytesIO(b"%b" % content.encode("utf8")),
             len(content),
         )
 
-    def file_changed(self, filename, bucket=S3_BUCKET):
-        """Check if file has changed since last call.
+    def watch(self, prefix: str, events=EventType.CREATED | EventType.REMOVED):
+        s3_events = []
+        if EventType.CREATED in events:
+            s3_events.append("s3:ObjectCreated:*")
+        if EventType.REMOVED in events:
+            s3_events.append("s3:ObjectRemoved:*")
+
+        with self.client.listen_bucket_notification(
+            self.bucket,
+            prefix=prefix,
+            events=s3_events,
+        ) as events:
+            for event in events:
+                yield Event(event)
+
+    def watch_async(
+        self,
+        prefix: str,
+        cb: callable,
+        events=EventType.CREATED | EventType.REMOVED,
+    ):
+        return Watcher(self, prefix, cb, events)
 
-        Arguments:
-            filename -- Local filename that is used
 
-        Keyword arguments:
-            bucket -- Storage bucket that is considered (default "seguro")
+class Watcher(threading.Thread):
+    def __init__(
+        self, client: Client, prefix: str, cb: callable, events: EventType
```

Comment:
Do I understand correctly that with that implementation we need individual `watcher` instances for each folder, i.e. prefix, we want to watch? Is that reasonable?

Reply:
Yes, unfortunately thats the case as the minio Python package does not support asynchronous IO natively.
In its core its doing a blocking read on the HTTP socket. Hence we must create a new watcher (OS thread) per prefix.

If we want to get fancy, we should use an async IO compatible minio package like this one: https://github.com/hlf20010508/miniopy-async

Reply:
Ok, but I think that is okay for the moment as we can still watch an entire prefix. When defining the directory hierarchy of the store, we should keep that mind as it influences how convenient we can track files.

Reply:
Minio also supports watching for a prefix + suffix match.
However, I didnt expose this option via the API yet